packages/cli/src/databases/repositories/test-case-execution.repository.ee.ts

```diff
@@ -0,0 +1,88 @@
+import { Service } from '@n8n/di';
+import { DataSource, In, Not, Repository } from '@n8n/typeorm';
+import type { DeepPartial } from '@n8n/typeorm/common/DeepPartial';
+
+import { TestCaseExecution } from '@/databases/entities/test-case-execution.ee';
+
+@Service()
+export class TestCaseExecutionRepository extends Repository<TestCaseExecution> {
+	constructor(dataSource: DataSource) {
+		super(TestCaseExecution, dataSource.manager);
+	}
+
+	async createBatch(testRunId: string, pastExecutionIds: string[]) {
+		const mappings = this.create(
+			pastExecutionIds.map<DeepPartial<TestCaseExecution>>((id) => ({
+				testRun: {
+					id: testRunId,
+				},
+				pastExecution: {
+					id,
+				},
+				status: 'new',
+			})),
+		);
+
+		return await this.save(mappings);
+	}
+
+	async markAsRunning(testRunId: string, pastExecutionId: string, executionId: string) {
+		return await this.update(
+			{ testRun: { id: testRunId }, pastExecutionId },
+			{
+				status: 'running',
+				executionId,
+				runAt: new Date(),
+			},
+		);
+	}
+
+	async markAsEvaluationRunning(
+		testRunId: string,
+		pastExecutionId: string,
+		evaluationExecutionId: string,
+	) {
+		return await this.update(
+			{ testRun: { id: testRunId }, pastExecutionId },
+			{
+				status: 'evaluation_running',
+				evaluationExecutionId,
+			},
+		);
+	}
+
+	async markAsCompleted(
+		testRunId: string,
+		pastExecutionId: string,
+		metrics: Record<string, number>,
+	) {
+		return await this.update(
+			{ testRun: { id: testRunId }, pastExecutionId },
+			{
+				status: 'success',
+				completedAt: new Date(),
+				metrics,
+			},
+		);
+	}
+
+	async markPendingAsCancelled(testRunId: string) {
+		return await this.update(
+			{ testRun: { id: testRunId }, status: Not(In(['success', 'error', 'cancelled'])) },
```

Comment:
Is it intentional that this is no-op (in contrast to throwing for example) if the status is not correct?

Reply:
The idea is when test run is interrupted by an instance crash or restart – we do not continue them after restart. This method is used to mark all test runs in an incomplete states as cancelled on instance startup.

Reply:
Check. Maybe name this `markAllPendingAsCancelled` to indicate it that it touches potentially multiple rows?

Reply:
I misguided you with my previous comment (deleted it). Good name suggestion though, changed it, thanks!